### ISSUE TITLE
improving weights for spherical area integrals/averages

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#633]
**Reported by Andrew Wittenberg on 4 Oct 2005 22:04 UTC**
After work today, I discovered something that may improve the computation
of areas in Ferret:

   http://mathworld.wolfram.com/Zone.html

With a little trigonometry, one can show that the ratio of the
Ferret-computed cell area over the true cell area is

   1/abs(sinc(yh/2))

where yh is the meridional cell width in radians.  This ratio depends only
on yh, and always exceeds 1.  This means that

1) Ferret's area-integrals are always a bit too large.

2) Ferret's area-averages, while exact over regions of uniform yh
(regularly-spaced latitudes), assign a bit too much weight to the more
coarsely-gridded latitudes when averaging over a non-uniform y-grid.

We can plot Ferret's fractional "area error" as a function of yh:

!----------------------------------------------
let pi = 3.141592653
let rad_per_deg = pi/180

let log_yhd = y[width"/unit="degrees_latitude" yhd = 10^log_yhd
let yh = rad_per_deg*yhd

let ratio = (yh/2)/abs(sin(yh/2))
let/title="fractional error in Ferret cell area" pct_err = ratio-1

plot/vs/line/hlog/vlog/hlim=1e-1:1e2/vlim=1e-7:1e-1 yhd, pct_err
!----------------------------------------------

The fractional error is below machine precision for yh<0.1deg, and is
fairly small (1e-4) even at 3deg meridional spacing.  But on coarse grids
with irregular latitude spacing, these errors might actually matter for
sensitive comparisons of global area-integrals & averages.

Here's a Ferret script to illustrate:

!----------------------------------------------
let nx = 2
let ny = 2
def ax/edges/x=0:360/unit=degrees_east/np=`nx` xax
def ax/edges/y=-90:90/unit=degrees_north/np=`ny` yax
let v = 1 + 0*x[gx=xax](gy=-1:2:.1]
let/title="ybox) + 0*y[pi = 3.141592653
let rad_per_deg = pi/180
let/title="radius of earth"/unit="m" r_earth = 6378100
let xh = rad_per_deg * xbox[gx=v](gy=yax]

let)
let yh = rad_per_deg * ybox[yc = rad_per_deg * y[gy=v](gy=v]
let)

let/title="actual area"/unit="m^2" area_actual = \\
   r_earth^2 * xh * cos(yc) * 2*abs(sin(yh/2))

! Note that this can also be expressed as:
!  r_earth^2 * xh * abs(sin(yc+yh/2)-sin(yc-yh/2))
!  r_earth^2 * xh * cos(yc) * yh * abs(sin(yh/2))/(yh/2)

let/title="Ferret area"/unit="m^2" area_ferret = \\                               
   r_earth^2 * xh * cos(yc) * yh                                                 
                                                                                 
list/prec=10 area_ferret[/ (4*pi*r_earth^2)                       
list/prec=10 area_actual[x=@sum,y=@sum](x=@sum,y=@sum]) / (4*pi*r_earth^2)                       
!----------------------------------------------                                  
                                                                                 
                                                                                 
Ansley & Steve, this little problem could easily be fixed inside of Ferret --    
instead of using a cos(yc) weighting, use a                                      
   cos(yc) * abs(sin(yh/2))/(yh/2)                                               
weighting.                                                                       
                                                                                 
Or for simplicity and computational efficiency, just replace yh with             
   2*abs(sin(yh/2))                                                              
                                                                                 
Cheers,                                                                          
                                                                                 
Andrew

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1348

**Comment by karlmsmith** [comment]
**Comment by Ansley.B.Manke@noaa.gov on 16 Mar 2006 22:39 UTC**
Made this correction in 

fer/doo/do_4d_trans.F
fer/doo/xyave_regrid.F
fer/plt/re_orient.F

**Comment by karlmsmith** [comment]
**Comment by Andrew.Wittenberg@noaa.gov on 17 May 2006 21:45 UTC**
The area of a quadrant of a spherical Earth is 

   (4*pi*r^2)/4 = pi*r^2 = 1.2780e14 m^2

Earlier Ferret versions gave an 11% error in the quadrant area:

        NOAA/PMEL TMAP
        FERRET v5.70
        IRIX 6.5 - 06/24/04
        17-May-06 17:27

yes? def ax/edges/x=0:360:180/unit=degrees_east xax
yes? def ax/edges/y=-90:90:90/unit=degrees_north yax
yes? let a = 1+0*x[list/prec=10 a[x=@din,y=@din](gx=xax]+0*y[gy=yax]
yes?)/4
             VARIABLE : A[            LONGITUDE: 0E to 0E(360)
             LATITUDE : 90S to 90N
          1.416347674E+14

The fixes in v6.00beta were meant to address this, but there's a problem:

        NOAA/PMEL TMAP
        FERRET v6
        Linux(g77) 2.4.21-32 - 04/26/06
        17-May-06 17:27

yes? def ax/edges/x=0:360:180/unit=degrees_east xax
yes? def ax/edges/y=-90:90:90/unit=degrees_north yax
yes? let a = 1+0*x[gx=xax](X=@DIN,Y=@DIN]/4
)+0*y[list/prec=10 a[x=@din,y=@din](gy=yax]
yes?)/4
             VARIABLE : A[X=@DIN,Y=@DIN]/4
             LONGITUDE: 0E to 0E(360)
             LATITUDE : 90S to 90N
          1246213504.

Is there a missing conversion from degrees to meters somewhere?

**Comment by karlmsmith** [comment]
**Comment by Ansley.B.Manke@noaa.gov on 6 Jun 2006 20:15 UTC**
There was indeed an error, essentially incorrect order of operations, applying
this correction and the units conversion degrees to meters in do_4d_trans.F

This is fixed, and a benchmark script added which tests the @DIN transorm and
several other transforms which had no benchmark tests. 

The code in xyave_regrid.F and re_orient.F is correct.

**Comment by karlmsmith** [comment]
**Comment by Andrew.Wittenberg@noaa.gov on 25 Aug 2006 22:22 UTC**
As of the official v6.00 release, Ferret's answer of 1.275161026E+14 m^2 for a
quadrant of the earth is now correct.

Earlier in this thread, I had mistakenly stated a slightly (0.2%) larger value
for the quadrant area.  This is because I had assumed an earth radius of
6378100m, which corresponds to the earth's circumference along the equator.  But
the earth is an "oblate spheroid" which bulges at the equator, so for area
calculations it's better to use the slightly smaller radius of 6371005m, the
"authalic mean radius," which is the "radius of a hypothetical perfect sphere
which has the same ... surface area as the sphereoid":

   http://en.wikipedia.org/wiki/Earth_radius

So Ferret v6.00 is right.  Hooray!

**Comment by karlmsmith** [comment]
**Comment by Ansley.B.Manke@noaa.gov on 4 Dec 2006 21:40 UTC**
The VECTOR/ASPECT command shows that the cosine correction is not made correctly:

USE monthly_navy_winds
vector/x=100:150/y=20:70/L=15 uwnd,vwnd
vector/x=100:150/y=20:70/L=15/asp/over uwnd,vwnd

**Comment by karlmsmith** [comment]
**Comment by Ansley.B.Manke@noaa.gov on 6 Apr 2007 21:44 UTC**
vector/aspect bug listed at the end of this is fixed; see the benchmark test
err600_vec_aspect.jnl

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 17 Aug 2009 21:39 UTC**
The above fix turns out not to be complete, as demonstrated by Marco Steinacher in his 8/17/2009 email to the Ferret User's Group.  The formula

```
cos(yc) * 2*abs(sin(yh/2))
```

is only identical to the more general formula

```
abs(sin(yhi)-sin(ylo))
```

when yc lies exactly midway between ylo and yhi.  Since area calculations really ought to be independent of where the cell center is located within a grid cell, can we (carefully) replace the first formula with the second?

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 8 Sep 2009 17:44 UTC**
This is fixed, with the correction made for Averages, Integrals, and the orientation of vectors as noted in comment above.  

Coordinate points may lie anywhere in the grid cell, starting with Ferret v6.0, and so this correction is needed.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 14 Jan 2010 23:27 UTC**
Area-weighted averages are still incorrect, whenever the latitudinal averaging limits don't sit exactly at cell boundaries.  To recap, based on

   [surface area of a lon/lat cell on the spherical Earth is computed as

```
let/title="authalic mean radius of Earth"/unit="m" r = 6371005
let pi = 3.1415926535897
let area = r^2 * (xhi-xlo)*pi/180 * abs(sin(yhi*pi/180)-sin(ylo*pi/180))
```

where xlo, xhi, ylo, yhi are the lon/lat limits of the cell in degrees.

I am attaching a script "area_bug.jnl," where we average over parts of two cells.  We get the wrong area weightings whenever the latitudinal averaging limits don't sit exactly at the cell boundaries, except when both the cells and the averaging limits are exactly symmetric about the equator.  The error scales with the distance between the averaging limits and the nearest cell bounds, so it should be small for fine grids, but can be large for coarse grids.

I would suggest adding these calls to area_bug.jnl (and any other end-cases you can think of) to the regression test suite for Ferret.

```
        NOAA/PMEL TMAP
        FERRET v6.401  
        Linux(g77) 2.4.21-57.ELsmp - 12/09/09
        14-Jan-10 18:09     

yes? go area_bug -40 40 -80 0 80  !THIS IS OK
             LONGITUDE: 0 to 60
             LATITUDE: -40 to 40
 Column  1: A1 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YBMID*PI/180) - SIN(YLO*PI/180))
 Column  2: A2 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YHI*PI/180) - SIN(YBMID*PI/180))
 Column  3: TRUE_AVE is (V1*A1 + V2*A2)/(A1+A2)
 Column  4: FER_AVE is F[X=0:60@AVE,Y=-40:40@AVE](http://mathworld.wolfram.com/Zone.html]

the)
                 A1      A2     TRUE_AV FER_AVE
I / *:     2.732E+13  2.732E+13  0.0000  0.0000
yes? 
yes? go area_bug -80 40 -80 0 40  !AND THIS IS OK
Replacing definition of axis YAX
             LONGITUDE: 0 to 60
             LATITUDE: -80 to 40
 Column  1: A1 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YBMID*PI/180) - SIN(YLO*PI/180))
 Column  2: A2 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YHI*PI/180) - SIN(YBMID*PI/180))
 Column  3: TRUE_AVE is (V1*A1 + V2*A2)/(A1+A2)
 Column  4: FER_AVE is F[                A1      A2     TRUE_AV FER_AVE
I / *:     4.186E+13  2.732E+13  -21.01  -21.01
yes? 
yes? go area_bug -40 80 -80 0 80  !BUT THIS IS INCORRECT
Replacing definition of axis YAX
             LONGITUDE: 0 to 60
             LATITUDE: -40 to 80
 Column  1: A1 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YBMID*PI/180) - SIN(YLO*PI/180))
 Column  2: A2 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YHI*PI/180) - SIN(YBMID*PI/180))
 Column  3: TRUE_AVE is (V1*A1 + V2*A2)/(A1+A2)
 Column  4: FER_AVE is F[X=0:60@AVE,Y=-40:80@AVE](X=0:60@AVE,Y=-80:40@AVE]
)
                 A1      A2     TRUE_AV FER_AVE
I / *:     2.732E+13  4.186E+13   21.01   33.33
yes? 
yes? go area_bug -40 40 -80 0 40  !ALSO INCORRECT
Replacing definition of axis YAX
             LONGITUDE: 0 to 60
             LATITUDE: -40 to 40
 Column  1: A1 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YBMID*PI/180) - SIN(YLO*PI/180))
 Column  2: A2 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YHI*PI/180) - SIN(YBMID*PI/180))
 Column  3: TRUE_AVE is (V1*A1 + V2*A2)/(A1+A2)
 Column  4: FER_AVE is F[                A1      A2     TRUE_AV FER_AVE
I / *:     2.732E+13  2.732E+13  0.0000   13.25
yes? 
yes? go area_bug -60 40 -80 0 40  !AND THIS IS INCORRECT TOO    
             LONGITUDE: 0 to 60
             LATITUDE: -60 to 40
 Column  1: A1 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YBMID*PI/180) - SIN(YLO*PI/180))
 Column  2: A2 is R^2 * (XBHI-XBLO)*PI/180 * ABS(SIN(YHI*PI/180) - SIN(YBMID*PI/180))
 Column  3: TRUE_AVE is (V1*A1 + V2*A2)/(A1+A2)
 Column  4: FER_AVE is F[X=0:60@AVE,Y=-60:40@AVE](X=0:60@AVE,Y=-40:40@AVE]
)
                 A1      A2     TRUE_AV FER_AVE
I / *:     3.681E+13  2.732E+13  -14.80  -6.936

```

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 15 Jan 2010 17:01 UTC**
I'm looking into this. The disagreement of the variables TRUE_AVE and FER_AVE computing area in this example script has always been present - the value of FER_AVE changes between v5.81 and v6.2 and then again between v6.2 and v6.3 but in all versions they still disagree widely.

The correct formula for the cosine correction is in the averaging code, perhaps the bug is in getting the partial box-sizes correct, and this particularly shows up with these large grid cells.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 26 Jan 2010 19:58 UTC**
This is fixed, and equivalent fix made to the integral @DIN.

Andrew, the 32-bit linux executable with this fix is at

ftp://ftp.pmel.noaa.gov/ferret/pub/linux_32/fer_executables.v642.tar.gz

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 4 Feb 2011 21:49 UTC**
We've found another serious bug related to cell areas, which seems to have first appeared in Ferret v6.42.  This was identified in a user's script that was trying to do some basic calculations of poleward ocean transports.

Ferret is giving badly incorrect answers when performing integrals along XZ and not over Y.  I haven't tested other axis/integral combinations, but would strongly recommend that you do so, and add those tests to the Ferret regression test suite.  The following script demonstrates the issue:

```
! Compute the correct area of a 10m-deep, 1-deglon cell at 26.5N.
let pi = 3.1415926
let/unit="m"/title="authalic radius of earth" r_authalic = 6371005
let/unit="degrees_north" lat = 26.5
let/unit="degrees_east" dx = 1 
let/unit="m" dz = 10
let dx_meters = r_authalic * cos(lat * pi/180) * (dx * pi/180)
list dx_meters*dz

! Next try to reproduce this with Ferret.
! First make a one-cell XYZ variable, with a value of 1.
def ax/x=279:280:1/modulo=360/unit=degrees_east/edge xax
def ax/y=26:27:1/unit=degrees_north/edge yax
def ax/z=0:10:10/unit=m/edge/depth zax
let p = 1+0*(x[Compute its integral (area) over XZ.
list p[x=279:380@din,z=0:10@din](gx=xax]+y[gy=yax]+z[gz=zax])

!)
```

For the above script, Ferret version 6.401 computes the correct answer for the second listing (995,121 m2).

Versions 6.42-6.6 give a missing value.

Versions 6.61-6.65 give a badly wrong answer (17,368 m2).

We are returning to version 6.401 until this is resolved.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 7 Feb 2011 20:46 UTC**
This is fixed. For an integral over one-point Y axis the corrections for latitude in this ticket, the calculation was off by a factor of 180/pi.

Gave Andrew an executable, which is the last official release, v6.65, with this one fix; so it is v6.651, GFDL-release only. The fix will also be in the next Ferret official release.

**Comment by karlmsmith** [comment]
**Attachment from andrew.wittenberg on 14 Jan 2010 23:28 UTC**
area_bug.jnl
REPLACE THIS TEXT WITH UPLOADED FILE ./attachments/TRAC_1348_GIT_633/area_bug.jnl
